emmalgs/all-my-friends
# Complete Intern Homework

**What this is**
Intern Homework
* A home page with an invite your friends button
* A modal appears with a list of invite candidates
* Send invites button is disabled until user selects candidate(s)
* Send invites button text updates as user selects and deselects candidates
* When clicked, modal closes, a post is sent to server, and a message appears with server response
* X button closes the modal

**Set Up**
* clone this repo
* in the parent directory, create `.env` file and paste the following:
```
REACT_APP_APICODE=[API key goes here]
REACT_APP_APIURL=https://olio-interview-homework.herokuapp.com
```
* in the parent directory in your terminal run `npm install`
* once all dependencies are installed run `npm run start`

**Media**
[Loom video of the app in action!](https://www.loom.com/share/95a54042f9e844b4a38825d7aa2d8df4?sid=90864237-cec3-4784-b831-3ef84c665363)
![homescreen](https://github.com/emmalgs/all-my-friends/assets/40610781/fcef5387-75fc-451b-a85c-30449ad51b58)
![modal with selection](https://github.com/emmalgs/all-my-friends/assets/40610781/f6502d26-34b7-42f6-8a00-35d5545a1acc)
![modal no selection](https://github.com/emmalgs/all-my-friends/assets/40610781/32cc1e0d-2214-4042-b2e7-1e676fa8e005)
![response message](https://github.com/emmalgs/all-my-friends/assets/40610781/9b6a4348-b082-42f8-8e31-2e5ee439ff73)

**Steps to Verify**
* In the parent directory in your terminal run `npm test` to verify tests are passing
![Screenshot 2023-09-26 at 12 39 51 PM](https://github.com/emmalgs/all-my-friends/assets/40610781/2bf62cd5-ef72-4dda-b526-599b7d33ea11)
* Once dependencies are installed and you run `npm run start` in the root directory, verify the following:
1. When `Invite Friends` is clicked, a modal appears
2. The button for `Send invites` is disabled if no friends are selected
3. If you select friends (the entire input field should also display a pointer and clicking the checkbox or name should make the selection) the button text updates based on the number of friends.
4. When you click `Send invites to x friends`, the home screen appears with the success message `Invites have been sent to x friends`.
5. At any point, if you click the x button in the modal, this returns you to the home screen.

**NOTES**
UI Structure & Logic:

```
src
	|__components
		|__atoms
		|	|__never actually used these atoms...
		|__molecules
		|	|__modalForm.jsx
		|	|__modalHeader.jsx
		|	|__selection.jsx
		|__screens
			|__home.jsx
			|__inviteController.jsx
			|__inviteFriendsModal.jsx
```

* I struggled with atoms, and realized what I was building were more molecules, hence, the lack of atoms in the atoms folder...
* the InviteController controls state for both modal and home, since it is in charge of fetching the friends list on render as well as the post method. This decides whether the home page is displaying or the modal is displaying, and what message is being displayed on the home page.
* Logic for border highlight when an email is selected is in the Selection molecule and uses state
* Logic for button disabled is in the inviteFriendsModal file instead of the modalForm file since the inviteFriendsModal tracks selected items. With atomic design, I'm still unsure where logic should exist, and I guess since this is such a small application it made sense to me to put it in the screen instead of the molecule.

Business Structure & Logic
```
src 
	|__ __tests__/api/inviteApi.test.js
	|__ services/inviteApi.js
```

* There is a services folder with the inviteAPI service. 
	* This is a class that when instantiated contains a base url and the api key
* In the `__tests__/api` folder I tested the inviteApi class by using Jest fetchMock
	* I originally built an express router to provide the same outputs as the api, however I ran into some out-of-my-pay-range concurrency issues and decided to go with fetchMock instead.
	* I tested error responses, but I have no idea if that is just redundant when using fetchMock?
